### PR TITLE
feat(provider/aws): Fix lifecycle hooks racing condition

### PR DESCRIPTION
**Pull request by costimuraru**
Fixes https://github.com/spinnaker/spinnaker/issues/5001

Below you can find some Deploy stage logs, which proves that lifecycle hooks are created before scaling out the server group:
```
{ 
   "phase":"AWS_DEPLOY",
   "status":"Beginning Amazon deployment."
},
{ 
   "phase":"AWS_DEPLOY",
   "status":"Beginning ASG deployment."
},
{ 
   "phase":"AWS_DEPLOY",
   "status":"Found ancestor server group, parsing details (name: testaws-stage-v009)"
},
{ 
   "phase":"AWS_DEPLOY",
   "status":"Deploying ASG: testaws-stage-v010"
},
{ 
   "phase":"AWS_DEPLOY",
   "status":" > Deploying to subnetIds: subnet-3d7dd13b,subnet-28b24012,subnet-5bc30543"
},
{ 
   "phase":"AWS_DEPLOY",
   "status":"Creating lifecycle hooks for: testaws-stage-v010"
},
{ 
   "phase":"AWS_DEPLOY",
   "status":"Creating lifecycle hook ({LifecycleHookName: testaws-stage-v010-lifecycle-8b39f839-6c5c-490c-ae3a-58518c1914b5,AutoScalingGroupName: testaws-stage-v010,LifecycleTransition: autoscaling:EC2_INSTANCE_LAUNCHING
},
{ 
   "phase":"AWS_DEPLOY",
   "status":"Creating lifecycle hook ({LifecycleHookName: testaws-stage-v010-lifecycle-08d2b7f7-d8a9-47dd-9157-7d86ee3a81d9,AutoScalingGroupName: testaws-stage-v010,LifecycleTransition: autoscaling:EC2_INSTANCE_TERMINATING
},
{ 
   "phase":"AWS_DEPLOY",
   "status":"Setting size of testaws-stage-v010 in aam-npe/us-east-1 to [min=1, max=1, desired=1]"
},
{ 
   "phase":"DEPLOY",
   "status":"Created the following deployments: [us-east-1:testaws-stage-v010]"
},
{ 
   "phase":"DEPLOY",
   "status":"Server Groups: [us-east-1:testaws-stage-v010] created."
},
{ 
   "phase":"ORCHESTRATION",
   "status":"Orchestration completed."
}
```

Previous logs were something like this (hooks created after scaling out):
```
...
{
"phase": "AWS_DEPLOY",
"status": "Setting size of recommservice-stage-va6-v286 in aam-npe/us-east-1 to [min=1, max=1, desired=1]"
},
{
"phase": "AWS_DEPLOY",
"status": "Creating lifecycle hook ({LifecycleHookName: recommservice-stage-va6-v286-lifecycle-2de43220-adb3-4151-8d81-8f2eb48f1944,AutoScalingGroupName: recommservice-stage-va6-v286,LifecycleTransition: autoscaling:EC2_INSTANCE_LAUNCHING"
},
{
"phase": "AWS_DEPLOY",
"status": "Creating lifecycle hook ({LifecycleHookName: recommservice-stage-va6-v286-lifecycle-ad1c18e1-c573-4914-bb3f-477946bdc067,AutoScalingGroupName: recommservice-stage-va6-v286,LifecycleTransition: autoscaling:EC2_INSTANCE_TERMINATING"
}
```

### CODE REVIEW

**Comment by costimuraru** [comment]
@robzienert @ajordens @dreynaud @cfieber please have a look at this small but meaningful bug fix.

This fixes the issue caused by the fact that the auto scale group is scaled out **before** the lifecycle hooks are attached to the ASG. This opens the door for a race condition, where an instance is sometimes created before the lifecycle hooks are attached.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/AutoScalingWorker.groovy [hunk 1]
```diff
@@ -267,6 +269,13 @@ class AutoScalingWorker {
       throw ex
     }
 
+    if (lifecycleHooks != null && lifecycleHooks.size() > 0) {
```
Comment (robzienert) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/AutoScalingWorker.groovy]:
```suggestion
    if (lifecycleHooks != null && !lifecycleHooks.isEmpty()) {
```

Prefer `isEmpty` over `size` for this case.

Comment (costimuraru) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/AutoScalingWorker.groovy]: Done @robzienert, thanks for the suggestion.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/BasicAmazonDeployHandler.groovy [hunk 2]
```diff
@@ -455,24 +454,6 @@ class BasicAmazonDeployHandler implements DeployHandler<BasicAmazonDeployDescrip
...
-                            BasicAmazonDeployDescription description,
-                            String targetAsgName) {
-
-    try {
-      List<AmazonAsgLifecycleHook> lifecycleHooks = getLifecycleHooks(targetCredentials, description)
-      if (lifecycleHooks.size() > 0) {
-        targetRegionScopedProvider.asgLifecycleHookWorker.attach(task, lifecycleHooks, targetAsgName)
-      }
-    } catch (Exception e) {
-      task.updateStatus(BASE_PHASE, "Unable to attach lifecycle hooks to ASG ($targetAsgName): ${e.message}")
```
Comment (dreynaud) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/BasicAmazonDeployHandler.groovy]: we used to log the exception message when attaching failed but not anymore?

Comment (costimuraru) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/BasicAmazonDeployHandler.groovy]: Good catch, @dreynaud! Re-added the error. Please have a look and let me know.

**Comment by costimuraru** [comment]
Thanks, @dreynaud @robzienert 
Let me know if we can merge this.

**Pull request merged**